scalding-core/src/main/scala/com/twitter/scalding/RichPipe.scala

```diff
@@ -95,6 +95,68 @@ object RichPipe extends java.io.Serializable {
     p
   }
 
+  /**
+   * Return true if a pipe is a source Pipe (has no parents / previous) and isn't a
+   * Splice.
+   */
+  def isSourcePipe(pipe: Pipe): Boolean = {
+    pipe.getParent == null &&
+      (pipe.getPrevious == null || pipe.getPrevious.isEmpty) &&
+      (!pipe.isInstanceOf[Splice])
+  }
+
+  def getPreviousPipe(p: Pipe): Option[Pipe] = {
+    if (p.getPrevious != null && p.getPrevious.length == 1) p.getPrevious.headOption
+    else None
+  }
+
+  /*
+   * If p1 represents a hashjoin, this method checks if
+   * p2 is one of the two sides in that hashjoin.
+   */
+  def isHashJoinedWithPipe(p1: Pipe, p2: Pipe): Boolean = {
```

Comment:
rename p1 to hashJoinPipe and p2 to hashJoinOperandPipe (or something on those lines?). Tends to be hard to keep track of p1, p2..